clickhouse_sqlalchemy/types.py

```diff
@@ -95,6 +95,9 @@ class Enum8(types.Enum):
     __visit_name__ = 'enum8'
 
     def __init__(self, *enums, **kw):
+        if not enums:
```

Comment:
Why do we need this?

Reply:
In some scenario the Enum instantiation was called from SQLAlchemy without any positional arguments, and with actual enums provided via `_enums` keyword argument.

Will try to find the original stacktrace to provide more reasoning with this particular change (if remembered correctly it was somehow related to reflection) -- perhaps I could move this two-lines of change as a separate pull request then.

Reply:
Ah, I've actually added the `test_enum` tests which trigger this as well. To answer the original question on why I added this change, see the following trace log -- there the instantiation is done without enums as positional arguments.

```
======================================================================
ERROR: test_enum8 (tests.test_reflection.ReflectionTestCase)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/clickhouse-sqlalchemy/tests/test_reflection.py", line 55, in test_enum8
    types.Enum8(enum.Enum('any8_enum', enum_options))
  File "/clickhouse-sqlalchemy/tests/test_reflection.py", line 19, in _type_round_trip
    with self.create_table(table):
  File "/usr/lib/python3.6/contextlib.py", line 81, in __enter__
    return next(self.gen)
  File "/clickhouse-sqlalchemy/tests/testcase.py", line 62, in create_table
    table.create(bind=self.session.bind)
  File "/clickhouse-sqlalchemy/venv/lib/python3.6/site-packages/sqlalchemy/sql/schema.py", line 779, in create
    checkfirst=checkfirst)
  File "/clickhouse-sqlalchemy/venv/lib/python3.6/site-packages/sqlalchemy/engine/base.py", line 1940, in _run_visitor
    conn._run_visitor(visitorcallable, element, **kwargs)
  File "/clickhouse-sqlalchemy/venv/lib/python3.6/site-packages/sqlalchemy/engine/base.py", line 1549, in _run_visitor
    **kwargs).traverse_single(element)
  File "/clickhouse-sqlalchemy/venv/lib/python3.6/site-packages/sqlalchemy/sql/visitors.py", line 121, in traverse_single
    return meth(obj, **kw)
  File "/clickhouse-sqlalchemy/venv/lib/python3.6/site-packages/sqlalchemy/sql/ddl.py", line 778, in visit_table
    _is_metadata_operation=_is_metadata_operation)
  File "/clickhouse-sqlalchemy/venv/lib/python3.6/site-packages/sqlalchemy/event/attr.py", line 284, in __call__
    fn(*args, **kw)
  File "/clickhouse-sqlalchemy/venv/lib/python3.6/site-packages/sqlalchemy/util/langhelpers.py", line 559, in __call__
    return getattr(self.target, self.name)(*arg, **kw)
  File "/clickhouse-sqlalchemy/venv/lib/python3.6/site-packages/sqlalchemy/sql/sqltypes.py", line 1093, in _on_table_create
    t = self.dialect_impl(bind.dialect)
  File "/clickhouse-sqlalchemy/venv/lib/python3.6/site-packages/sqlalchemy/sql/type_api.py", line 446, in dialect_impl
    return self._dialect_info(dialect)['impl']
  File "/clickhouse-sqlalchemy/venv/lib/python3.6/site-packages/sqlalchemy/sql/type_api.py", line 510, in _dialect_info
    impl = self.adapt(type(self))
  File "/clickhouse-sqlalchemy/venv/lib/python3.6/site-packages/sqlalchemy/sql/sqltypes.py", line 1451, in adapt
    return super(Enum, self).adapt(impltype, **kw)
  File "/clickhouse-sqlalchemy/venv/lib/python3.6/site-packages/sqlalchemy/sql/type_api.py", line 729, in adapt
    return self.adapt_to_emulated(impltype, **kw)
  File "/clickhouse-sqlalchemy/venv/lib/python3.6/site-packages/sqlalchemy/sql/sqltypes.py", line 1447, in adapt_to_emulated
    return impltype(**kw)
  File "/clickhouse-sqlalchemy/clickhouse_sqlalchemy/types.py", line 103, in __init__
    self.enum_type = enums[0]
IndexError: tuple index out of range
-------------------- >> begin captured logging << --------------------
clickhouse_driver.connection: DEBUG: Query: 
DROP TABLE IF EXISTS t
clickhouse_driver.connection: DEBUG: Block send time: 0.000240
--------------------- >> end captured logging << ---------------------
```

Reply:
Well, here is the same code in SA: https://github.com/zzzeek/sqlalchemy/blob/35a1f5481a28837490037a6bd99d63590f748933/lib/sqlalchemy/sql/sqltypes.py#L1338

This was introduced in 1.2.0 version. I was comparing with older one.

Good job!